apm-agent-api/src/main/java/co/elastic/apm/api/ElasticApm.java

```diff
@@ -1,40 +1,121 @@
 package co.elastic.apm.api;
 
+/**
+ * This class can be used to statically access the {@link Tracer}.
+ * <p>
+ * You can store the Tracer as an instance like so:
+ * </p>
+ * <pre>{@code
+ * private static final Tracer tracer = ElasticApm.get();
+ * }</pre>
+ * <p>
+ * Then you access the tracer to set a custom transaction name,
+ * for example:
+ * <pre>{@code
+ * tracer.currentTransaction().setName("SuchController#muchMethod");
+ * }</pre>
+ * <p/>
+ */
 public class ElasticApm implements Tracer {
 
+    /**
+     * The singleton instance of the Tracer.
+     */
     private static final ElasticApm INSTANCE = new ElasticApm();
 
     private static Tracer tracer = NoopTracer.INSTANCE;
```

Comment:
I think this variable should not be static, you are inside a singleton. See below.

About thread safety, there are several parts where you access this variable in a non thread safe and some other ie. register/unregister where you have a`synchonized` block, that makes this class not thread safety.

Adding `synchronized` to all the methods will add too much contention here I think that the safest / most performant option is to make this variable volatile.

You could instead add a `ReadWrite` lock, but that seems overkill in this situation and make everything a lot more complex.

Reply:
Then I'd have to make the `INSTANCE` variable package protected in order to be able to access the instance methods `register` and `unregister`, but that seems like a good option.

Sure, I could use volatile in order to make sure that I'm always reading the current value of `tracer` in each thread. But the worst case implication of an outdated read in this case is only that the `NoopTracer` is returned. I'd argue that is not really that important, as it only done once at startup time. Furthermore, ElasticAPM is initialized at the very start of the application. That means that it is unlikely that a Thread has ever read the tracer variable and thus being in a CPU cache.

Reply:
@danielmitterdorfer could you shed some light on wether a non-volatile read can lead to reading a corrupted value or if the worst case is only reading a slightly outdated value?

Reply:
In the general case it can indeed happen that you read an inconsistent object state (to be clear: the object *reference* is always consistent (see [JLS §17.7](https://docs.oracle.com/javase/specs/jls/se9/html/jls-17.html#jls-17.7). In this special case, you are publishing this field via a static initializer which is always thread-safe (according to [JLS §12.4.2](https://docs.oracle.com/javase/specs/jls/se9/html/jls-12.html#jls-12.4.2)) so you should be fine here. However, IMHO you need to be careful with republication (i.e. assigning a different tracer instance to that field). It depends a bit on the tracer implementation whether you can do this in a thread-safe manner (e.g. if the tracer instance is immutable then even racy publishing is safe).